### ISSUE TITLE
OSX Terminal sometimes loses CWD info when some processes are launched

### ISSUE THREAD

**Issue by elia** [fish-shell/fish-shell#2418]
Alas I still have to find a way to reproduce reliably, but happens several times a day

**Comment by ridiculousfish** [comment]
I don't understand. What does "loses CWD info" mean?

**Comment by faho** [comment]
@elia: Ping? Could you clarify what you meant here?

**Comment by elia** [comment]
_Sorry, missed the first notification._

So, what happens is this:

1. Open a new Terminal.app
2. I `cd ~/Code/my-rails-app`
3. start a process (e.g. `./bin/rails console`)
4. hit <kbd>⌘T</kbd> to open a new tab

**Expected**

- the working directory of the new tab should be already in `~/Code/my-rails-app`

**Actual**

- the working directory of the new tab is `~`

--- 

I'm aware there's a special escape sequence that the shell sends to the terminal to notify it about the current directory but of course I'm not sure what is going wrong… by the randomness I'd say that looks like some threading issue where the CWD is erroneously taken form another terminal window instead of the current one. But that's just a hunch, nothing more 😊

**Comment by faho** [comment]
This is likely to be influenced (maybe even caused) by code in config.fish (and related) - mind posting that?

**Comment by elia** [comment]
config.fish

```fish
if status --is-login
  set PATH $PATH ~/bin
end

alias g=git
export EDITOR="mate -w"
export GEM_EDITOR="mate"

function fish_prompt
  set git_dir (git rev-parse --show-toplevel ^ /dev/null)
  if test -n "$git_dir"
    # set_color cyan
    # echo -n (~/.rvm/bin/rvm-exec ~/.rvm/bin/rvm-prompt u)" "
    set_color aaa
    echo -n (basename $git_dir)
    set_color 444
    __fish_git_prompt ":%s "
  else
    # set_color aaa
    # echo -n (prompt_pwd)
  end

  echo -n '✨  '
  set_color normal
end

function fish_right_prompt -d "Write out the right prompt"
  set_color aaa
  echo -n (prompt_pwd)
  set_color normal
end

ram > /dev/null
```

functions/rvm.fish

```
function rvm --description='Ruby enVironment Manager'
  # run RVM and capture the resulting environment
  set --local env_file (mktemp -t rvm.fish.XXXXXXXXXX)
  bash -c 'source ~/.rvm/scripts/rvm; rvm "$@"; status=$?; env > "$0"; exit $status' $env_file $argv

  # apply rvm_* and *PATH variables from the captured environment
  and eval (grep '^rvm\|^[^=]*PATH\|^GEM_HOME' $env_file | grep -v '_clr=' | sed '/^[^=]*PATH/s/:/" "/g; s/^/set -xg /; s/=/ "/; s/$/" ;/; s/(//; s/)//')
  # clean up
  rm -f $env_file
end

function __handle_rvmrc_stuff --on-variable PWD
  # Source a .rvmrc file in a directory after changing to it, if it exists.
  # To disable this fature, set rvm_project_rvmrc=0 in $HOME/.rvmrc
  if test "$rvm_project_rvmrc" != 0
    set -l cwd $PWD
    while true
      if contains $cwd "" $HOME "/"
        if test "$rvm_project_rvmrc_default" = 1
          rvm default 1>/dev/null 2>&1
        end
        break
      else
        if test -e .rvmrc -o -e .ruby-version -o -e .ruby-gemset
          eval "rvm reload" > /dev/null
          eval "rvm rvmrc load" >/dev/null
          break
        else
          set cwd (dirname "$cwd")
        end
      end
    end

    set -e cwd
  end
end
```

**Comment by faho** [comment]
>ram > /dev/null

Did you mean `rvm`?

I can't see any probable cause here. @ridiculousfish: Your patient.

**Comment by elia** [comment]
@faho just blame OS X autocorrection :disappointed:

**Comment by ridiculousfish** [comment]
My memory is that fish doesn't inform Terminal about the cwd. bash does this in `/etc/bashrc` on OS X.

**Comment by elia** [comment]
Good to know, maybe Terminal.app has some heuristic that fails occasionally. Trying to implement the following Terminal.app.instructions:

```
The working directory may be set using the Operating System Command (OSC) escape sequence:

	ESC ] Ps ; Pt BEL

The parameter Ps is 7 (working directory) and Pt is a “file:” URL. The URL should include a hostname to disambiguate local and remote paths, and characters must be percent-encoded as appropriate.
```

with this function:

```fish
function set_osx_pwd
  echo -ne "\e]7;file://"(ruby -rcgi -e'CGI.escape "'(hostname)'//"+ENV["PWD"]')"/\a"` 
end
```

and adding it to to `fish_prompt`, let me know if it can be improved somehow. 

--- 

Of course will report back if that makes the issue disappear.

**Comment by krader1961** [comment]
> My memory is that fish doesn't inform Terminal about the cwd. bash does this in /etc/bashrc on OS X.

Specifically, it's done via the /etc/bashrc_Apple_Terminal script which modifies PROMPT_COMMAND to run a function named update_terminal_cwd defined in that script. That's a script written and distributed by Apple -- it's not even part of the standard bash distribution AFAICT.

I'm a longtime fan of Mac OS X and want as much support for it as possible but fail to see why this should be considered a fish issue. @elia did the solution you mentioned in your Sep 28 comment result in the behavior you expect? It should based on my reading of the Apple supplied bashrc_Apple_Terminal script. Also, iTerm2 (which I recommend over Apple Terminal) doesn't support that escape sequence. Obviously that isn't a fish issue :-) This might warrant a FAQ entry but a solution shouldn't be a part of the core fish software.

**Comment by zanchey** [comment]
> My memory is that fish doesn't inform Terminal about the cwd

It does - see #68 & #906.

**Comment by krader1961** [comment]
Okay, __fish_config_interactive.fish contains the following block of code which would seemingly implement the desired functionality:

```
        # Notify vte-based terminals when $PWD changes (issue #906)
        if test "$VTE_VERSION" -ge 3405 -o "$TERM_PROGRAM" = "Apple_Terminal"
                function __update_vte_cwd --on-variable PWD --description 'Notify VTE of change to $
                        status --is-command-substitution; and return
                        printf '\033]7;file://%s%s\a' (hostname) (pwd | __fish_urlencode)
                end
        end
```

It works for me when I launch Terminal.app rather than iTerm.app.

I think this portion of the O.P. problem statement is relevant:

> ...I'd say that looks like some threading issue where the CWD is erroneously taken form another terminal window instead of the current one.

The nature of this mechanism is such that race conditions and unexpected behavior are guaranteed. Especially in the face of a power-user who might be executing functions in an interactive shell that changes PWD.

@elia what happens if you `cd ~/Code/my-rails-app` then open a new tab? Do you consistently get a fish shell in the new tab with the expected CWD?

**Comment by floam** [comment]
I notice on initially opening a terminal.app window, it doesn't really report having a path URI until after the first directory change. Checking the inspector, until I cd $pwd the "Directory" location isn't initialized with something. Or just run printf 'file://%s%s\a' (hostname) (pwd | __fish_urlencode) once.

**Comment by faho** [comment]
Is this as simple as just running `__update_vte_cwd` once, unconditionally?

**Comment by floam** [comment]
Yes.

**Comment by krader1961** [comment]
> Is this as simple as just running __update_vte_cwd once, unconditionally?

As @floam said, the answer is "yes". If I simply enter `__update_vte_cwd ` it corrects the information in the terminal title bar. Also, just entering `cd .` corrects the PWD reported by terminal.app. It should also be noted that the CWD is correct when opening a new tab; "correct" meaning that it is in fact the PWD of the previous tab, it is just the terminal.app title that is incorrect. So the solution would seem to be to execute `__update_vte_cwd ` immediately after it is defined in `__fish_config_interactive.fish`.

**Comment by faho** [comment]
Please test my attempt at a fix - I can't since I can't reproduce this with any vte-terminal and don't have OSX anywhere.

**Comment by floam** [comment]
Exactly that same change to HEAD fixed the starting behavior on OS X 10.11.2 for me.

**Comment by elia** [comment]
Thank very much for the fix! Looks good so far :rocket:
